File: 238 Product of Array Except Self/Python/Solution.py
from typing import List, Self


class Solution:
    def productExceptSelf(self: Self, nums: List[int]) -> List[int]:
        size: int = len(nums)
        result: List[int] = [1] * size
            
        i: int = 0
        multiple: int = 1
        for j in range(1, size):
            multiple *= nums[i]
            result[j] *= multiple
            i += 1
            
        i = size - 1
        multiple = 1
        for j in range(size - 2, -1, -1):
            multiple *= nums[i]
            result[j] *= multiple
            i -= 1
            
        return result

File: 238 Product of Array Except Self/C/solution.c
#include <stdlib.h>
#include <string.h>

#include "solution.h"

int* productExceptSelf(int* nums, int numsSize, int* returnSize) {
    *returnSize = numsSize;
    int* result = (int*) malloc(sizeof(int) * *returnSize);
    for (int i = 0; i < *returnSize; i++) {
        result[i] = 1;
    }

    int i = 0;
    int multiple = 1;
    for (int j = 1; j < numsSize; j++) {
        multiple *= nums[i];
        result[j] *= multiple;
        i++;
    }

    i = numsSize - 1;
    multiple = 1;
    for (int j = numsSize - 2; j >= 0; j--) {
        multiple *= nums[i];
        result[j] *= multiple;
        i--;
    }

    return result;
}